// storyblok-js-client/dist/schema.mjs
var i = Object.defineProperty;
var g = (e, r, t) => r in e ? i(e, r, { enumerable: !0, configurable: !0, writable: !0, value: t }) : e[r] = t;
var a = (e, r, t) => g(e, typeof r != "symbol" ? r + "" : r, t);
class p {
  constructor() {
    a(this, "isCDNUrl", (r = "") => r.includes("/cdn/"));
    a(this, "getOptionsPage", (r, t = 25, n = 1) => ({
      ...r,
      per_page: t,
      page: n
    }));
    a(this, "delay", (r) => new Promise((t) => setTimeout(t, r)));
    a(this, "arrayFrom", (r = 0, t) => Array.from({ length: r }, t));
    a(this, "range", (r = 0, t = r) => {
      const n = Math.abs(t - r) || 0, s = r < t ? 1 : -1;
      return this.arrayFrom(n, (o, c) => c * s + r);
    });
    a(this, "asyncMap", async (r, t) => Promise.all(r.map(t)));
    a(this, "flatMap", (r = [], t) => r.map(t).reduce((n, s) => [...n, ...s], []));
    /**
     * @method escapeHTML
     * @param  {string} string text to be parsed
     * @return {string} Text parsed
     */
    a(this, "escapeHTML", function(r) {
      const t = {
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': "&quot;",
        "'": "&#39;"
      }, n = /[&<>"']/g, s = new RegExp(n.source);
      return r && s.test(r) ? r.replace(n, (o) => t[o]) : r;
    });
  }
  /**
   * @method stringify
   * @param  {object} params
   * @param  {string} prefix
   * @param  {boolean} isArray
   * @return {string} Stringified object
   */
  stringify(r, t, n) {
    const s = [];
    for (const o in r) {
      if (!Object.prototype.hasOwnProperty.call(r, o))
        continue;
      const c = r[o], l = n ? "" : encodeURIComponent(o);
      let u;
      typeof c == "object" ? u = this.stringify(
        c,
        t ? t + encodeURIComponent(`[${l}]`) : l,
        Array.isArray(c)
      ) : u = `${t ? t + encodeURIComponent(`[${l}]`) : l}=${encodeURIComponent(c)}`, s.push(u);
    }
    return s.join("&");
  }
  /**
   * @method getRegionURL
   * @param  {string} regionCode region code, could be eu, us, cn, ap or ca
   * @return {string} The base URL of the region
   */
  getRegionURL(r) {
    const t = "api.storyblok.com", n = "api-us.storyblok.com", s = "app.storyblokchina.cn", o = "api-ap.storyblok.com", c = "api-ca.storyblok.com";
    switch (r) {
      case "us":
        return n;
      case "cn":
        return s;
      case "ap":
        return o;
      case "ca":
        return c;
      default:
        return t;
    }
  }
}
const m = function(e, r) {
  const t = {};
  for (const n in e) {
    const s = e[n];
    r.includes(n) && s !== null && (t[n] = s);
  }
  return t;
}, h = (e) => e === "email", y = () => ({
  singleTag: "hr"
}), f = () => ({
  tag: "blockquote"
}), d = () => ({
  tag: "ul"
}), k = (e) => ({
  tag: [
    "pre",
    {
      tag: "code",
      attrs: e.attrs
    }
  ]
}), b = () => ({
  singleTag: "br"
}), _ = (e) => ({
  tag: `h${e.attrs.level}`
}), U = (e) => ({
  singleTag: [
    {
      tag: "img",
      attrs: m(e.attrs, ["src", "alt", "title"])
    }
  ]
}), R = () => ({
  tag: "li"
}), A = () => ({
  tag: "ol"
}), L = () => ({
  tag: "p"
}), P = (e) => ({
  tag: [
    {
      tag: "span",
      attrs: {
        "data-type": "emoji",
        "data-name": e.attrs.name,
        emoji: e.attrs.emoji
      }
    }
  ]
}), $ = () => ({
  tag: "b"
}), I = () => ({
  tag: "s"
}), T = () => ({
  tag: "u"
}), j = () => ({
  tag: "strong"
}), C = () => ({
  tag: "code"
}), H = () => ({
  tag: "i"
}), M = (e) => {
  if (!e.attrs)
    return {
      tag: ""
    };
  const r = new p().escapeHTML, t = { ...e.attrs }, { linktype: n = "url" } = e.attrs;
  if (delete t.linktype, t.href && (t.href = r(e.attrs.href || "")), h(n) && (t.href = `mailto:${t.href}`), t.anchor && (t.href = `${t.href}#${t.anchor}`, delete t.anchor), t.custom) {
    for (const s in t.custom)
      t[s] = t.custom[s];
    delete t.custom;
  }
  return {
    tag: [
      {
        tag: "a",
        attrs: t
      }
    ]
  };
}, w = (e) => ({
  tag: [
    {
      tag: "span",
      attrs: e.attrs
    }
  ]
}), E = () => ({
  tag: "sub"
}), q = () => ({
  tag: "sup"
}), v = (e) => ({
  tag: [
    {
      tag: "span",
      attrs: e.attrs
    }
  ]
}), O = (e) => {
  var t;
  return (t = e.attrs) != null && t.color ? {
    tag: [
      {
        tag: "span",
        attrs: {
          style: `background-color:${e.attrs.color};`
        }
      }
    ]
  } : {
    tag: ""
  };
}, S = (e) => {
  var t;
  return (t = e.attrs) != null && t.color ? {
    tag: [
      {
        tag: "span",
        attrs: {
          style: `color:${e.attrs.color}`
        }
      }
    ]
  } : {
    tag: ""
  };
}, N = {
  nodes: {
    horizontal_rule: y,
    blockquote: f,
    bullet_list: d,
    code_block: k,
    hard_break: b,
    heading: _,
    image: U,
    list_item: R,
    ordered_list: A,
    paragraph: L,
    emoji: P
  },
  marks: {
    bold: $,
    strike: I,
    underline: T,
    strong: j,
    code: C,
    italic: H,
    link: M,
    styled: w,
    subscript: E,
    superscript: q,
    anchor: v,
    highlight: O,
    textStyle: S
  }
};
export {
  N as default
};
