// @marvr/storyblok-rich-text-types/dist/storyblok-rich-text-types.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = global || self, factory(global.StoryblokRichTextTypes = {}));
}(this, (function (exports) { 'use strict';

  (function (Block) {
      Block["DOCUMENT"] = "doc";
      Block["HEADING"] = "heading";
      Block["PARAGRAPH"] = "paragraph";
      Block["QUOTE"] = "blockquote";
      Block["OL_LIST"] = "ordered_list";
      Block["UL_LIST"] = "bullet_list";
      Block["LIST_ITEM"] = "list_item";
      Block["CODE"] = "code_block";
      Block["HR"] = "horizontal_rule";
      Block["BR"] = "hard_break";
      Block["IMAGE"] = "image";
  })(exports.Block || (exports.Block = {}));

  (function (Mark) {
      Mark["BOLD"] = "bold";
      Mark["STRONG"] = "strong";
      Mark["STRIKE"] = "strike";
      Mark["UNDERLINE"] = "underline";
      Mark["ITALIC"] = "italic";
      Mark["CODE"] = "code";
      Mark["LINK"] = "link";
      Mark["STYLED"] = "styled";
  })(exports.Mark || (exports.Mark = {}));

  const ComponentType = 'blok';

  const TextType = 'text';

  exports.ComponentType = ComponentType;
  exports.TextType = TextType;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
