// @storyblok/richtext/dist/storyblok-richtext.es.js
/**
 * name: @storyblok/richtext
 * (c) 2025
 * description: Storyblok RichText Resolver
 * author: Alvaro Saburido <hola@alvarosaburido.dev> (https://github.com/alvarosabu/)
 */
function q(e, n) {
  if (!n)
    return { src: e, attrs: {} };
  let R = 0, S = 0;
  const E = {}, i = [];
  function v(h, u, $, L, I) {
    typeof h != "number" || h <= u || h >= $ ? console.warn(`[StoryblokRichText] - ${L.charAt(0).toUpperCase() + L.slice(1)} value must be a number between ${u} and ${$} (inclusive)`) : I.push(`${L}(${h})`);
  }
  if (typeof n == "object") {
    if (typeof n.width == "number" && n.width > 0 ? (E.width = n.width, R = n.width) : console.warn("[StoryblokRichText] - Width value must be a number greater than 0"), n.height && typeof n.height == "number" && n.height > 0 ? (E.height = n.height, S = n.height) : console.warn("[StoryblokRichText] - Height value must be a number greater than 0"), n.loading && ["lazy", "eager"].includes(n.loading) && (E.loading = n.loading), n.class && (E.class = n.class), n.filters) {
      const { filters: h } = n || {}, { blur: u, brightness: $, fill: L, format: I, grayscale: _, quality: m, rotate: O } = h || {};
      u && v(u, 0, 100, "blur", i), m && v(m, 0, 100, "quality", i), $ && v($, 0, 100, "brightness", i), L && i.push(`fill(${L})`), _ && i.push("grayscale()"), O && [0, 90, 180, 270].includes(n.filters.rotate || 0) && i.push(`rotate(${O})`), I && ["webp", "png", "jpeg"].includes(I) && i.push(`format(${I})`);
    }
    n.srcset && (E.srcset = n.srcset.map((h) => {
      if (typeof h == "number")
        return `${e}/m/${h}x0/${i.length > 0 ? `filters:${i.join(":")}` : ""} ${h}w`;
      if (Array.isArray(h) && h.length === 2) {
        const [u, $] = h;
        return `${e}/m/${u}x${$}/${i.length > 0 ? `filters:${i.join(":")}` : ""} ${u}w`;
      } else {
        console.warn("[StoryblokRichText] - srcset entry must be a number or a tuple of two numbers");
        return;
      }
    }).join(", ")), n.sizes && (E.sizes = n.sizes.join(", "));
  }
  let A = `${e}/m/`;
  return R > 0 && S > 0 && (A = `${A}${R}x${S}/`), i.length > 0 && (A = `${A}filters:${i.join(":")}`), {
    src: A,
    attrs: E
  };
}
var o = /* @__PURE__ */ ((e) => (e.DOCUMENT = "doc", e.HEADING = "heading", e.PARAGRAPH = "paragraph", e.QUOTE = "blockquote", e.OL_LIST = "ordered_list", e.UL_LIST = "bullet_list", e.LIST_ITEM = "list_item", e.CODE_BLOCK = "code_block", e.HR = "horizontal_rule", e.BR = "hard_break", e.IMAGE = "image", e.EMOJI = "emoji", e.COMPONENT = "blok", e.TABLE = "table", e.TABLE_ROW = "tableRow", e.TABLE_CELL = "tableCell", e.TABLE_HEADER = "tableHeader", e))(o || {}), g = /* @__PURE__ */ ((e) => (e.BOLD = "bold", e.STRONG = "strong", e.STRIKE = "strike", e.UNDERLINE = "underline", e.ITALIC = "italic", e.CODE = "code", e.LINK = "link", e.ANCHOR = "anchor", e.STYLED = "styled", e.SUPERSCRIPT = "superscript", e.SUBSCRIPT = "subscript", e.TEXT_STYLE = "textStyle", e.HIGHLIGHT = "highlight", e))(g || {}), N = /* @__PURE__ */ ((e) => (e.TEXT = "text", e))(N || {}), k = /* @__PURE__ */ ((e) => (e.SELF = "_self", e.BLANK = "_blank", e))(k || {}), w = /* @__PURE__ */ ((e) => (e.URL = "url", e.STORY = "story", e.ASSET = "asset", e.EMAIL = "email", e))(w || {});
const B = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
], W = (e = {}) => Object.keys(e).map((n) => `${n}="${e[n]}"`).join(" "), X = (e = {}) => Object.keys(e).map((n) => `${n}: ${e[n]}`).join("; ");
function J(e) {
  return e.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
const T = (e) => Object.fromEntries(Object.entries(e).filter(([n, R]) => R !== void 0));
function p(e, n = {}, R) {
  const S = W(n), E = S ? `${e} ${S}` : e, i = Array.isArray(R) ? R.join("") : R || "";
  if (e) {
    if (B.includes(e))
      return `<${E}>`;
  } else return i;
  return `<${E}>${i}</${e}>`;
}
function M(e = {}) {
  const n = /* @__PURE__ */ new Map(), {
    renderFn: R = p,
    textFn: S = J,
    resolvers: E = {},
    optimizeImages: i = !1,
    keyedResolvers: v = !1
  } = e, A = R !== p, h = () => ({ render: (r, s = {}, a) => {
    if (v && r) {
      const l = n.get(r) || 0;
      n.set(r, l + 1), s.key = `${r}-${l}`;
    }
    return R(r, s, a);
  } }), u = (t) => (r, s) => {
    const a = r.attrs || {};
    return s.render(t, a, r.children || null);
  }, $ = (t, r) => {
    const { src: s, alt: a, title: l, srcset: d, sizes: c } = t.attrs || {};
    let f = s, b = {};
    if (i) {
      const { src: K, attrs: Y } = q(s, i);
      f = K, b = Y;
    }
    const F = {
      src: f,
      alt: a,
      title: l,
      srcset: d,
      sizes: c,
      ...b
    };
    return r.render("img", T(F));
  }, L = (t, r) => {
    const { level: s, ...a } = t.attrs || {};
    return r.render(`h${s}`, a, t.children);
  }, I = (t, r) => {
    var a, l, d, c;
    const s = r.render("img", {
      src: (a = t.attrs) == null ? void 0 : a.fallbackImage,
      alt: (l = t.attrs) == null ? void 0 : l.alt,
      style: "width: 1.25em; height: 1.25em; vertical-align: text-top",
      draggable: "false",
      loading: "lazy"
    });
    return r.render("span", {
      "data-type": "emoji",
      "data-name": (d = t.attrs) == null ? void 0 : d.name,
      "data-emoji": (c = t.attrs) == null ? void 0 : c.emoji
    }, s);
  }, _ = (t, r) => r.render(
    "pre",
    t.attrs || {},
    r.render("code", {}, t.children || "")
  ), m = (t, r = !1) => ({ text: s, attrs: a }, l) => {
    const { class: d, id: c, ...f } = a || {}, b = r ? {
      class: d,
      id: c,
      style: X(f) || void 0
    } : a || {};
    return l.render(t, T(b), s);
  }, O = (t) => x(t), H = (t) => {
    const { marks: r, ...s } = t;
    if ("text" in t) {
      if (r)
        return r.reduce(
          (l, d) => O({ ...d, text: l }),
          O({ ...s, children: s.children })
        );
      const a = t.attrs || {};
      if (v) {
        const l = n.get("txt") || 0;
        n.set("txt", l + 1), a.key = `txt-${l}`;
      }
      return S(s.text, a);
    }
    return "";
  }, j = (t, r) => {
    const { linktype: s, href: a, anchor: l, ...d } = t.attrs || {};
    let c = "";
    switch (s) {
      case w.ASSET:
      case w.URL:
        c = a;
        break;
      case w.EMAIL:
        c = `mailto:${a}`;
        break;
      case w.STORY:
        c = a, l && (c = `${c}#${l}`);
        break;
      default:
        c = a;
        break;
    }
    const f = { ...d };
    return c && (f.href = c), r.render("a", f, t.text);
  }, y = (t, r) => {
    var s, a;
    return console.warn("[StoryblokRichtText] - BLOK resolver is not available for vanilla usage"), r.render("span", {
      blok: (s = t == null ? void 0 : t.attrs) == null ? void 0 : s.body[0],
      id: (a = t.attrs) == null ? void 0 : a.id,
      style: "display: none"
    });
  }, D = (t, r) => {
    const s = {}, a = r.render("tbody", {}, t.children);
    return r.render("table", s, a);
  }, P = (t, r) => {
    const s = {};
    return r.render("tr", s, t.children);
  }, U = (t, r) => {
    const { colspan: s, rowspan: a, colwidth: l, backgroundColor: d, ...c } = t.attrs || {}, f = {
      ...c
    };
    s > 1 && (f.colspan = s), a > 1 && (f.rowspan = a);
    const b = [];
    return l && b.push(`width: ${l}px;`), d && b.push(`background-color: ${d};`), b.length > 0 && (f.style = b.join(" ")), r.render("td", T(f), t.children);
  }, G = (t, r) => {
    const { colspan: s, rowspan: a, colwidth: l, backgroundColor: d, ...c } = t.attrs || {}, f = {
      ...c
    };
    s > 1 && (f.colspan = s), a > 1 && (f.rowspan = a);
    const b = [];
    return l && b.push(`width: ${l}px;`), d && b.push(`background-color: ${d};`), b.length > 0 && (f.style = b.join(" ")), r.render("th", T(f), t.children);
  }, z = new Map([
    [o.DOCUMENT, u("")],
    [o.HEADING, L],
    [o.PARAGRAPH, u("p")],
    [o.UL_LIST, u("ul")],
    [o.OL_LIST, u("ol")],
    [o.LIST_ITEM, u("li")],
    [o.IMAGE, $],
    [o.EMOJI, I],
    [o.CODE_BLOCK, _],
    [o.HR, u("hr")],
    [o.BR, u("br")],
    [o.QUOTE, u("blockquote")],
    [o.COMPONENT, y],
    [N.TEXT, H],
    [g.LINK, j],
    [g.ANCHOR, j],
    [g.STYLED, m("span", !0)],
    [g.BOLD, m("strong")],
    [g.TEXT_STYLE, m("span", !0)],
    [g.ITALIC, m("em")],
    [g.UNDERLINE, m("u")],
    [g.STRIKE, m("s")],
    [g.CODE, m("code")],
    [g.SUPERSCRIPT, m("sup")],
    [g.SUBSCRIPT, m("sub")],
    [g.HIGHLIGHT, m("mark")],
    [o.TABLE, D],
    [o.TABLE_ROW, P],
    [o.TABLE_CELL, U],
    [o.TABLE_HEADER, G],
    ...Object.entries(E).map(([t, r]) => [t, r])
  ]);
  function C(t) {
    const r = z.get(t.type);
    if (!r)
      return console.error("<Storyblok>", `No resolver found for node type ${t.type}`), "";
    const s = h();
    if (t.type === "text")
      return r(t, s);
    const a = t.content ? t.content.map(x) : void 0;
    return r({
      ...t,
      children: a
    }, s);
  }
  function x(t) {
    return t.type === "doc" ? A ? t.content.map(C) : t.content.map(C).join("") : Array.isArray(t) ? t.map(C) : C(t);
  }
  return {
    render: x
  };
}
export {
  o as BlockTypes,
  k as LinkTargets,
  w as LinkTypes,
  g as MarkTypes,
  N as TextTypes,
  M as richTextResolver
};
