// @storyblok/richtext/dist/storyblok-richtext.es.js
/**
 * name: @storyblok/richtext
 * (c) 2025
 * description: Storyblok RichText Resolver
 * author: Alvaro Saburido <hola@alvarosaburido.dev> (https://github.com/alvarosabu/)
 */
function K(t, r) {
  if (!r)
    return { src: t, attrs: {} };
  let i = 0, S = 0;
  const g = {}, o = [];
  function m(l, E, R, L, I) {
    typeof l != "number" || l <= E || l >= R ? console.warn(`[StoryblokRichText] - ${L.charAt(0).toUpperCase() + L.slice(1)} value must be a number between ${E} and ${R} (inclusive)`) : I.push(`${L}(${l})`);
  }
  if (typeof r == "object") {
    if (typeof r.width == "number" && r.width > 0 ? (g.width = r.width, i = r.width) : console.warn("[StoryblokRichText] - Width value must be a number greater than 0"), r.height && typeof r.height == "number" && r.height > 0 ? (g.height = r.height, S = r.height) : console.warn("[StoryblokRichText] - Height value must be a number greater than 0"), r.loading && ["lazy", "eager"].includes(r.loading) && (g.loading = r.loading), r.class && (g.class = r.class), r.filters) {
      const { filters: l } = r || {}, { blur: E, brightness: R, fill: L, format: I, grayscale: $, quality: v, rotate: w } = l || {};
      E && m(E, 0, 100, "blur", o), v && m(v, 0, 100, "quality", o), R && m(R, 0, 100, "brightness", o), L && o.push(`fill(${L})`), $ && o.push("grayscale()"), w && [0, 90, 180, 270].includes(r.filters.rotate || 0) && o.push(`rotate(${w})`), I && ["webp", "png", "jpeg"].includes(I) && o.push(`format(${I})`);
    }
    r.srcset && (g.srcset = r.srcset.map((l) => {
      if (typeof l == "number")
        return `${t}/m/${l}x0/${o.length > 0 ? `filters:${o.join(":")}` : ""} ${l}w`;
      if (Array.isArray(l) && l.length === 2) {
        const [E, R] = l;
        return `${t}/m/${E}x${R}/${o.length > 0 ? `filters:${o.join(":")}` : ""} ${E}w`;
      } else {
        console.warn("[StoryblokRichText] - srcset entry must be a number or a tuple of two numbers");
        return;
      }
    }).join(", ")), r.sizes && (g.sizes = r.sizes.join(", "));
  }
  let A = `${t}/m/`;
  return i > 0 && S > 0 && (A = `${A}${i}x${S}/`), o.length > 0 && (A = `${A}filters:${o.join(":")}`), {
    src: A,
    attrs: g
  };
}
var f = /* @__PURE__ */ ((t) => (t.DOCUMENT = "doc", t.HEADING = "heading", t.PARAGRAPH = "paragraph", t.QUOTE = "blockquote", t.OL_LIST = "ordered_list", t.UL_LIST = "bullet_list", t.LIST_ITEM = "list_item", t.CODE_BLOCK = "code_block", t.HR = "horizontal_rule", t.BR = "hard_break", t.IMAGE = "image", t.EMOJI = "emoji", t.COMPONENT = "blok", t.TABLE = "table", t.TABLE_ROW = "tableRow", t.TABLE_CELL = "tableCell", t.TABLE_HEADER = "tableHeader", t))(f || {}), d = /* @__PURE__ */ ((t) => (t.BOLD = "bold", t.STRONG = "strong", t.STRIKE = "strike", t.UNDERLINE = "underline", t.ITALIC = "italic", t.CODE = "code", t.LINK = "link", t.ANCHOR = "anchor", t.STYLED = "styled", t.SUPERSCRIPT = "superscript", t.SUBSCRIPT = "subscript", t.TEXT_STYLE = "textStyle", t.HIGHLIGHT = "highlight", t))(d || {}), p = /* @__PURE__ */ ((t) => (t.TEXT = "text", t))(p || {}), F = /* @__PURE__ */ ((t) => (t.SELF = "_self", t.BLANK = "_blank", t))(F || {}), O = /* @__PURE__ */ ((t) => (t.URL = "url", t.STORY = "story", t.ASSET = "asset", t.EMAIL = "email", t))(O || {});
const Y = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
], q = (t = {}) => Object.keys(t).map((r) => `${r}="${t[r]}"`).join(" "), W = (t = {}) => Object.keys(t).map((r) => `${r}: ${t[r]}`).join("; ");
function X(t) {
  return t.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
const _ = (t) => Object.fromEntries(Object.entries(t).filter(([r, i]) => i !== void 0));
function j(t, r = {}, i) {
  const S = q(r), g = S ? `${t} ${S}` : t, o = Array.isArray(i) ? i.join("") : i || "";
  if (t) {
    if (Y.includes(t))
      return `<${g}>`;
  } else return o;
  return `<${g}>${o}</${t}>`;
}
function B(t = {}) {
  let r = 0;
  const {
    renderFn: i = j,
    textFn: S = X,
    resolvers: g = {},
    optimizeImages: o = !1,
    keyedResolvers: m = !1
  } = t, A = i !== j, l = (e) => (s) => {
    const n = s.attrs || {};
    return m && (n.key = `${e}-${r}`), i(e, n, s.children || null);
  }, E = (e) => {
    const { src: s, alt: n, title: c, srcset: b, sizes: u } = e.attrs || {};
    let a = s, h = {};
    if (o) {
      const { src: G, attrs: z } = K(s, o);
      a = G, h = z;
    }
    m && (h = {
      ...h,
      key: `img-${r}`
    });
    const k = {
      src: a,
      alt: n,
      title: c,
      srcset: b,
      sizes: u,
      ...h
    };
    return i("img", _(k));
  }, R = (e) => {
    const { level: s, ...n } = e.attrs || {}, c = {
      ...n
    };
    return m && (c.key = `h${s}-${r}`), i(`h${s}`, c, e.children);
  }, L = (e) => {
    var c, b, u, a;
    const s = i("img", {
      src: (c = e.attrs) == null ? void 0 : c.fallbackImage,
      alt: (b = e.attrs) == null ? void 0 : b.alt,
      style: "width: 1.25em; height: 1.25em; vertical-align: text-top",
      draggable: "false",
      loading: "lazy"
    }), n = {
      "data-type": "emoji",
      "data-name": (u = e.attrs) == null ? void 0 : u.name,
      "data-emoji": (a = e.attrs) == null ? void 0 : a.emoji
    };
    return m && (n.key = `emoji-${r}`), i("span", n, s);
  }, I = (e) => i("pre", {
    ...e.attrs,
    key: `code-${r}`
  }, i("code", { key: `code-${r}` }, e.children || "")), $ = (e, s = !1) => ({ text: n, attrs: c }) => {
    const { class: b, id: u, ...a } = c || {}, h = s ? {
      class: b,
      id: u,
      style: W(a) || void 0
    } : c || {};
    return m && (h.key = `${e}-${r}`), i(e, _(h), n);
  }, v = (e) => y(e), w = (e) => {
    const { marks: s, ...n } = e;
    return "text" in e ? s ? s.reduce(
      (c, b) => v({ ...b, text: c }),
      // Fix: Ensure render function returns a string
      v({ ...n, children: n.children })
      // Fix: Cast children to string
    ) : S(n.text) : "";
  }, C = (e) => {
    const { linktype: s, href: n, anchor: c, ...b } = e.attrs || {};
    let u = "";
    switch (s) {
      case O.ASSET:
      case O.URL:
        u = n;
        break;
      case O.EMAIL:
        u = `mailto:${n}`;
        break;
      case O.STORY:
        u = n, c && (u = `${u}#${c}`);
        break;
      default:
        u = n;
        break;
    }
    const a = { ...b };
    return u && (a.href = u), m && (a.key = `a-${r}`), i("a", a, e.text);
  }, N = (e) => {
    var s, n;
    return console.warn("[StoryblokRichtText] - BLOK resolver is not available for vanilla usage"), i("span", {
      blok: (s = e == null ? void 0 : e.attrs) == null ? void 0 : s.body[0],
      id: (n = e.attrs) == null ? void 0 : n.id,
      key: `component-${r}`,
      style: "display: none"
    });
  }, H = (e) => {
    const s = {};
    return m && (s.key = `table-${r}`), i("table", s, e.children);
  }, x = (e) => {
    const s = {};
    return m && (s.key = `tr-${r}`), i("tr", s, e.children);
  }, D = (e) => {
    const { colspan: s, rowspan: n, colwidth: c, backgroundColor: b, ...u } = e.attrs || {}, a = {
      ...u
    };
    s > 1 && (a.colspan = s), n > 1 && (a.rowspan = n);
    const h = [];
    return c && h.push(`width: ${c}px;`), b && h.push(`background-color: ${b};`), h.length > 0 && (a.style = h.join(" ")), m && (a.key = `td-${r}`), i("td", _(a), e.children);
  }, P = (e) => {
    const { colspan: s, rowspan: n, colwidth: c, backgroundColor: b, ...u } = e.attrs || {}, a = {
      ...u
    };
    s > 1 && (a.colspan = s), n > 1 && (a.rowspan = n);
    const h = [];
    return c && h.push(`width: ${c}px;`), b && h.push(`background-color: ${b};`), h.length > 0 && (a.style = h.join(" ")), m && (a.key = `th-${r}`), i("th", _(a), e.children);
  }, U = new Map([
    [f.DOCUMENT, l("")],
    [f.HEADING, R],
    [f.PARAGRAPH, l("p")],
    [f.UL_LIST, l("ul")],
    [f.OL_LIST, l("ol")],
    [f.LIST_ITEM, l("li")],
    [f.IMAGE, E],
    [f.EMOJI, L],
    [f.CODE_BLOCK, I],
    [f.HR, l("hr")],
    [f.BR, l("br")],
    [f.QUOTE, l("blockquote")],
    [f.COMPONENT, N],
    [p.TEXT, w],
    [d.LINK, C],
    [d.ANCHOR, C],
    [d.STYLED, $("span", !0)],
    [d.BOLD, $("strong")],
    [d.TEXT_STYLE, $("span", !0)],
    [d.ITALIC, $("em")],
    [d.UNDERLINE, $("u")],
    [d.STRIKE, $("s")],
    [d.CODE, $("code")],
    [d.SUPERSCRIPT, $("sup")],
    [d.SUBSCRIPT, $("sub")],
    [d.HIGHLIGHT, $("mark")],
    [f.TABLE, H],
    [f.TABLE_ROW, x],
    [f.TABLE_CELL, D],
    [f.TABLE_HEADER, P],
    ...Object.entries(g).map(([e, s]) => [e, s])
  ]);
  function T(e) {
    r += 1;
    const s = U.get(e.type);
    if (!s)
      return console.error("<Storyblok>", `No resolver found for node type ${e.type}`), "";
    if (e.type === "text")
      return s(e);
    const n = e.content ? e.content.map(y) : void 0;
    return s({
      ...e,
      children: n
      // Fix: Update the type of 'children' to Node[]
    });
  }
  function y(e) {
    return e.type === "doc" ? A ? e.content.map(T) : e.content.map(T).join("") : Array.isArray(e) ? e.map(T) : T(e);
  }
  return {
    render: y
  };
}
export {
  f as BlockTypes,
  F as LinkTargets,
  O as LinkTypes,
  d as MarkTypes,
  p as TextTypes,
  B as richTextResolver
};
