// @storyblok/richtext/dist/storyblok-richtext.es.js
/**
 * name: @storyblok/richtext
 * (c) 2024
 * description: Storyblok RichText Resolver
 * author: Alvaro Saburido <hola@alvarosaburido.dev> (https://github.com/alvarosabu/)
 */
function U(e, r) {
  if (!r)
    return { src: e, attrs: {} };
  let a = 0, R = 0;
  const b = {}, l = [];
  function d(c, g, S, E, n) {
    typeof c != "number" || c <= g || c >= S ? console.warn(`[StoryblokRichText] - ${E.charAt(0).toUpperCase() + E.slice(1)} value must be a number between ${g} and ${S} (inclusive)`) : n.push(`${E}(${c})`);
  }
  if (typeof r == "object") {
    if (typeof r.width == "number" && r.width > 0 ? (b.width = r.width, a = r.width) : console.warn("[StoryblokRichText] - Width value must be a number greater than 0"), r.height && typeof r.height == "number" && r.height > 0 ? (b.height = r.height, R = r.height) : console.warn("[StoryblokRichText] - Height value must be a number greater than 0"), r.loading && ["lazy", "eager"].includes(r.loading) && (b.loading = r.loading), r.class && (b.class = r.class), r.filters) {
      const { filters: c } = r || {}, { blur: g, brightness: S, fill: E, format: n, grayscale: L, quality: T, rotate: v } = c || {};
      g && d(g, 0, 100, "blur", l), T && d(T, 0, 100, "quality", l), S && d(S, 0, 100, "brightness", l), E && l.push(`fill(${E})`), L && l.push("grayscale()"), v && [0, 90, 180, 270].includes(r.filters.rotate || 0) && l.push(`rotate(${v})`), n && ["webp", "png", "jpeg"].includes(n) && l.push(`format(${n})`);
    }
    r.srcset && (b.srcset = r.srcset.map((c) => {
      if (typeof c == "number")
        return `${e}/m/${c}x0/${l.length > 0 ? `filters:${l.join(":")}` : ""} ${c}w`;
      if (Array.isArray(c) && c.length === 2) {
        const [g, S] = c;
        return `${e}/m/${g}x${S}/${l.length > 0 ? `filters:${l.join(":")}` : ""} ${g}w`;
      } else {
        console.warn("[StoryblokRichText] - srcset entry must be a number or a tuple of two numbers");
        return;
      }
    }).join(", ")), r.sizes && (b.sizes = r.sizes.join(", "));
  }
  let u = `${e}/m/`;
  return a > 0 && R > 0 && (u = `${u}${a}x${R}/`), l.length > 0 && (u = `${u}filters:${l.join(":")}`), {
    src: u,
    attrs: b
  };
}
var f = /* @__PURE__ */ ((e) => (e.DOCUMENT = "doc", e.HEADING = "heading", e.PARAGRAPH = "paragraph", e.QUOTE = "blockquote", e.OL_LIST = "ordered_list", e.UL_LIST = "bullet_list", e.LIST_ITEM = "list_item", e.CODE_BLOCK = "code_block", e.HR = "horizontal_rule", e.BR = "hard_break", e.IMAGE = "image", e.EMOJI = "emoji", e.COMPONENT = "blok", e))(f || {}), m = /* @__PURE__ */ ((e) => (e.BOLD = "bold", e.STRONG = "strong", e.STRIKE = "strike", e.UNDERLINE = "underline", e.ITALIC = "italic", e.CODE = "code", e.LINK = "link", e.ANCHOR = "anchor", e.STYLED = "styled", e.SUPERSCRIPT = "superscript", e.SUBSCRIPT = "subscript", e.TEXT_STYLE = "textStyle", e.HIGHLIGHT = "highlight", e))(m || {}), _ = /* @__PURE__ */ ((e) => (e.TEXT = "text", e))(_ || {}), D = /* @__PURE__ */ ((e) => (e.SELF = "_self", e.BLANK = "_blank", e))(D || {}), A = /* @__PURE__ */ ((e) => (e.URL = "url", e.STORY = "story", e.ASSET = "asset", e.EMAIL = "email", e))(A || {});
const G = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
], z = (e = {}) => Object.keys(e).map((r) => `${r}="${e[r]}"`).join(" "), p = (e = {}) => Object.keys(e).map((r) => `${r}: ${e[r]}`).join("; ");
function k(e) {
  return e.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
const y = (e) => Object.fromEntries(Object.entries(e).filter(([r, a]) => a !== void 0));
function K(e, r = {}, a) {
  const R = z(r), b = R ? `${e} ${R}` : e;
  return G.includes(e) ? `<${b}>` : `<${b}>${Array.isArray(a) ? a.join("") : a || ""}</${e}>`;
}
function F(e = {}) {
  let r = 0;
  const {
    renderFn: a = K,
    textFn: R = k,
    resolvers: b = {},
    optimizeImages: l = !1,
    keyedResolvers: d = !1
  } = e, u = (t) => (i) => {
    const s = i.attrs || {};
    return d && (s.key = `${t}-${r}`), a(t, s, i.children || null);
  }, c = (t) => {
    const { src: i, alt: s, title: o, srcset: $, sizes: h } = t.attrs || {};
    let I = i, O = {};
    if (l) {
      const { src: H, attrs: P } = U(i, l);
      I = H, O = P;
    }
    d && (O = {
      ...O,
      key: `img-${r}`
    });
    const x = {
      src: I,
      alt: s,
      title: o,
      srcset: $,
      sizes: h,
      ...O
    };
    return a("img", y(x));
  }, g = (t) => {
    const { level: i, ...s } = t.attrs || {}, o = {
      ...s
    };
    return d && (o.key = `h${i}-${r}`), a(`h${i}`, o, t.children);
  }, S = (t) => {
    var o, $, h, I;
    const i = a("img", {
      src: (o = t.attrs) == null ? void 0 : o.fallbackImage,
      alt: ($ = t.attrs) == null ? void 0 : $.alt,
      style: "width: 1.25em; height: 1.25em; vertical-align: text-top",
      draggable: "false",
      loading: "lazy"
    }), s = {
      "data-type": "emoji",
      "data-name": (h = t.attrs) == null ? void 0 : h.name,
      "data-emoji": (I = t.attrs) == null ? void 0 : I.emoji
    };
    return d && (s.key = `emoji-${r}`), a("span", s, i);
  }, E = (t) => a("pre", {
    ...t.attrs,
    key: `code-${r}`
  }, a("code", { key: `code-${r}` }, t.children || "")), n = (t, i = !1) => ({ text: s, attrs: o }) => {
    const { class: $, id: h, ...I } = o || {}, O = i ? {
      class: $,
      id: h,
      style: p(I) || void 0
    } : o || {};
    return d && (O.key = `${t}-${r}`), a(t, y(O), s);
  }, L = (t) => N(t), T = (t) => {
    const { marks: i, ...s } = t;
    return "text" in t ? i ? i.reduce(
      (o, $) => L({ ...$, text: o }),
      // Fix: Ensure render function returns a string
      L({ ...s, children: s.children })
      // Fix: Cast children to string
    ) : R(s.text) : "";
  }, v = (t) => {
    const { linktype: i, href: s, anchor: o, ...$ } = t.attrs || {};
    let h = "";
    switch (i) {
      case A.ASSET:
      case A.URL:
        h = s;
        break;
      case A.EMAIL:
        h = `mailto:${s}`;
        break;
      case A.STORY:
        h = s;
        break;
    }
    return o && (h = `${h}#${o}`), a("a", { ...$, href: h, key: `a-${r}` }, t.text);
  }, w = (t) => {
    var i, s;
    return console.warn("[StoryblokRichtText] - BLOK resolver is not available for vanilla usage"), a("span", {
      blok: (i = t == null ? void 0 : t.attrs) == null ? void 0 : i.body[0],
      id: (s = t.attrs) == null ? void 0 : s.id,
      key: `component-${r}`,
      style: "display: none"
    });
  }, C = new Map([
    [f.DOCUMENT, u("div")],
    [f.HEADING, g],
    [f.PARAGRAPH, u("p")],
    [f.UL_LIST, u("ul")],
    [f.OL_LIST, u("ol")],
    [f.LIST_ITEM, u("li")],
    [f.IMAGE, c],
    [f.EMOJI, S],
    [f.CODE_BLOCK, E],
    [f.HR, u("hr")],
    [f.BR, u("br")],
    [f.QUOTE, u("blockquote")],
    [f.COMPONENT, w],
    [_.TEXT, T],
    [m.LINK, v],
    [m.ANCHOR, v],
    [m.STYLED, n("span", !0)],
    [m.BOLD, n("strong")],
    [m.TEXT_STYLE, n("span", !0)],
    [m.ITALIC, n("em")],
    [m.UNDERLINE, n("u")],
    [m.STRIKE, n("s")],
    [m.CODE, n("code")],
    [m.SUPERSCRIPT, n("sup")],
    [m.SUBSCRIPT, n("sub")],
    [m.HIGHLIGHT, n("mark")],
    ...Object.entries(b).map(([t, i]) => [t, i])
  ]);
  function j(t) {
    r += 1;
    const i = C.get(t.type);
    if (!i)
      return console.error("<Storyblok>", `No resolver found for node type ${t.type}`), "";
    if (t.type === "text")
      return i(t);
    const s = t.content ? t.content.map(N) : void 0;
    return i({
      ...t,
      children: s
      // Fix: Update the type of 'children' to Node[]
    });
  }
  function N(t) {
    return Array.isArray(t) ? t.map(j) : j(t);
  }
  return {
    render: N
  };
}
export {
  f as BlockTypes,
  D as LinkTargets,
  A as LinkTypes,
  m as MarkTypes,
  _ as TextTypes,
  F as richTextResolver
};
