// happy-dom/lib/exception/DOMExceptionNameEnum.js
var DOMExceptionNameEnum;
(function (DOMExceptionNameEnum) {
    DOMExceptionNameEnum["invalidStateError"] = "InvalidStateError";
    DOMExceptionNameEnum["indexSizeError"] = "IndexSizeError";
    DOMExceptionNameEnum["syntaxError"] = "SyntaxError";
    DOMExceptionNameEnum["hierarchyRequestError"] = "HierarchyRequestError";
    DOMExceptionNameEnum["notSupportedError"] = "NotSupportedError";
    DOMExceptionNameEnum["wrongDocumentError"] = "WrongDocumentError";
    DOMExceptionNameEnum["invalidNodeTypeError"] = "InvalidNodeTypeError";
    DOMExceptionNameEnum["invalidCharacterError"] = "InvalidCharacterError";
    DOMExceptionNameEnum["notFoundError"] = "NotFoundError";
    DOMExceptionNameEnum["securityError"] = "SecurityError";
    DOMExceptionNameEnum["networkError"] = "NetworkError";
    DOMExceptionNameEnum["domException"] = "DOMException";
    DOMExceptionNameEnum["invalidAccessError"] = "InvalidAccessError";
    DOMExceptionNameEnum["unknownError"] = "UnknownError";
    DOMExceptionNameEnum["abortError"] = "AbortError";
    DOMExceptionNameEnum["timeoutError"] = "TimeoutError";
    DOMExceptionNameEnum["encodingError"] = "EncodingError";
    DOMExceptionNameEnum["uriMismatchError"] = "URIMismatchError";
    DOMExceptionNameEnum["inUseAttributeError"] = "InUseAttributeError";
    DOMExceptionNameEnum["namespaceError"] = "NamespaceError";
})(DOMExceptionNameEnum || (DOMExceptionNameEnum = {}));
export default DOMExceptionNameEnum;
