// @storyblok/vue/dist/storyblok-vue.mjs
/**
 * name: @storyblok/vue
 * (c) 2025
 * description: SDK to integrate Storyblok into your project using Vue.
 * author: Storyblok
 */
import { defineComponent as J, ref as O, resolveDynamicComponent as H, inject as de, createBlock as Y, openBlock as W, mergeProps as pe, h as X, createTextVNode as ge, watch as fe, onMounted as me, defineAsyncComponent as ye } from "vue";
let z = !1;
const F = [], be = (r) => new Promise((e, t) => {
  if (typeof window > "u" || (window.storyblokRegisterEvent = (o) => {
    if (window.location === window.parent.location) {
      console.warn("You are not in Draft Mode or in the Visual Editor.");
      return;
    }
    z ? o() : F.push(o);
  }, document.getElementById("storyblok-javascript-bridge")))
    return;
  const s = document.createElement("script");
  s.async = !0, s.src = r, s.id = "storyblok-javascript-bridge", s.onerror = (o) => t(o), s.onload = (o) => {
    F.forEach((n) => n()), z = !0, e(o);
  }, document.getElementsByTagName("head")[0].appendChild(s);
});
var ke = Object.defineProperty, ve = (r, e, t) => e in r ? ke(r, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : r[e] = t, m = (r, e, t) => ve(r, typeof e != "symbol" ? e + "" : e, t);
class $e extends Error {
  constructor(e) {
    super(e), this.name = "AbortError";
  }
}
function Te(r, e, t) {
  if (!Number.isFinite(e))
    throw new TypeError("Expected `limit` to be a finite number");
  if (!Number.isFinite(t))
    throw new TypeError("Expected `interval` to be a finite number");
  const s = [];
  let o = [], n = 0, a = !1;
  const i = async () => {
    n++;
    const h = s.shift();
    if (h)
      try {
        const p = await r(...h.args);
        h.resolve(p);
      } catch (p) {
        h.reject(p);
      }
    const u = setTimeout(() => {
      n--, s.length > 0 && i(), o = o.filter((p) => p !== u);
    }, t);
    o.includes(u) || o.push(u);
  }, l = (...h) => a ? Promise.reject(
    new Error(
      "Throttled function is already aborted and not accepting new promises"
    )
  ) : new Promise((u, p) => {
    s.push({
      resolve: u,
      reject: p,
      args: h
    }), n < e && i();
  });
  return l.abort = () => {
    a = !0, o.forEach(clearTimeout), o = [], s.forEach(
      (h) => h.reject(() => new $e("Throttle function aborted"))
    ), s.length = 0;
  }, l;
}
class P {
  constructor() {
    m(this, "isCDNUrl", (e = "") => e.includes("/cdn/")), m(this, "getOptionsPage", (e, t = 25, s = 1) => ({
      ...e,
      per_page: t,
      page: s
    })), m(this, "delay", (e) => new Promise((t) => setTimeout(t, e))), m(this, "arrayFrom", (e = 0, t) => Array.from({ length: e }, t)), m(this, "range", (e = 0, t = e) => {
      const s = Math.abs(t - e) || 0, o = e < t ? 1 : -1;
      return this.arrayFrom(s, (n, a) => a * o + e);
    }), m(this, "asyncMap", async (e, t) => Promise.all(e.map(t))), m(this, "flatMap", (e = [], t) => e.map(t).reduce((s, o) => [...s, ...o], [])), m(this, "escapeHTML", function(e) {
      const t = {
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': "&quot;",
        "'": "&#39;"
      }, s = /[&<>"']/g, o = new RegExp(s.source);
      return e && o.test(e) ? e.replace(s, (n) => t[n]) : e;
    });
  }
  /**
   * @method stringify
   * @param  {object} params
   * @param  {string} prefix
   * @param  {boolean} isArray
   * @return {string} Stringified object
   */
  stringify(e, t, s) {
    const o = [];
    for (const n in e) {
      if (!Object.prototype.hasOwnProperty.call(e, n))
        continue;
      const a = e[n];
      if (a == null)
        continue;
      const i = s ? "" : encodeURIComponent(n);
      let l;
      typeof a == "object" ? l = this.stringify(
        a,
        t ? t + encodeURIComponent(`[${i}]`) : i,
        Array.isArray(a)
      ) : l = `${t ? t + encodeURIComponent(`[${i}]`) : i}=${encodeURIComponent(a)}`, o.push(l);
    }
    return o.join("&");
  }
  /**
   * @method getRegionURL
   * @param  {string} regionCode region code, could be eu, us, cn, ap or ca
   * @return {string} The base URL of the region
   */
  getRegionURL(e) {
    const t = "api.storyblok.com", s = "api-us.storyblok.com", o = "app.storyblokchina.cn", n = "api-ap.storyblok.com", a = "api-ca.storyblok.com";
    switch (e) {
      case "us":
        return s;
      case "cn":
        return o;
      case "ap":
        return n;
      case "ca":
        return a;
      default:
        return t;
    }
  }
}
const we = function(r, e) {
  const t = {};
  for (const s in r) {
    const o = r[s];
    e.includes(s) && o !== null && (t[s] = o);
  }
  return t;
}, Re = (r) => r === "email", _e = () => ({
  singleTag: "hr"
}), Se = () => ({
  tag: "blockquote"
}), Ee = () => ({
  tag: "ul"
}), xe = (r) => ({
  tag: [
    "pre",
    {
      tag: "code",
      attrs: r.attrs
    }
  ]
}), je = () => ({
  singleTag: "br"
}), Ae = (r) => ({
  tag: `h${r.attrs.level}`
}), Ie = (r) => ({
  singleTag: [
    {
      tag: "img",
      attrs: we(r.attrs, ["src", "alt", "title"])
    }
  ]
}), Ce = () => ({
  tag: "li"
}), Le = () => ({
  tag: "ol"
}), Oe = () => ({
  tag: "p"
}), Pe = (r) => ({
  tag: [
    {
      tag: "span",
      attrs: {
        "data-type": "emoji",
        "data-name": r.attrs.name,
        emoji: r.attrs.emoji
      }
    }
  ]
}), Ne = () => ({
  tag: "b"
}), Me = () => ({
  tag: "s"
}), He = () => ({
  tag: "u"
}), De = () => ({
  tag: "strong"
}), Ue = () => ({
  tag: "code"
}), Be = () => ({
  tag: "i"
}), ze = (r) => {
  if (!r.attrs)
    return {
      tag: ""
    };
  const e = new P().escapeHTML, t = { ...r.attrs }, { linktype: s = "url" } = r.attrs;
  if (delete t.linktype, t.href && (t.href = e(r.attrs.href || "")), Re(s) && (t.href = `mailto:${t.href}`), t.anchor && (t.href = `${t.href}#${t.anchor}`, delete t.anchor), t.custom) {
    for (const o in t.custom)
      t[o] = t.custom[o];
    delete t.custom;
  }
  return {
    tag: [
      {
        tag: "a",
        attrs: t
      }
    ]
  };
}, Fe = (r) => ({
  tag: [
    {
      tag: "span",
      attrs: r.attrs
    }
  ]
}), qe = () => ({
  tag: "sub"
}), Ve = () => ({
  tag: "sup"
}), Ge = (r) => ({
  tag: [
    {
      tag: "span",
      attrs: r.attrs
    }
  ]
}), Ke = (r) => {
  var e;
  return (e = r.attrs) != null && e.color ? {
    tag: [
      {
        tag: "span",
        attrs: {
          style: `background-color:${r.attrs.color};`
        }
      }
    ]
  } : {
    tag: ""
  };
}, Je = (r) => {
  var e;
  return (e = r.attrs) != null && e.color ? {
    tag: [
      {
        tag: "span",
        attrs: {
          style: `color:${r.attrs.color}`
        }
      }
    ]
  } : {
    tag: ""
  };
}, Ye = {
  nodes: {
    horizontal_rule: _e,
    blockquote: Se,
    bullet_list: Ee,
    code_block: xe,
    hard_break: je,
    heading: Ae,
    image: Ie,
    list_item: Ce,
    ordered_list: Le,
    paragraph: Oe,
    emoji: Pe
  },
  marks: {
    bold: Ne,
    strike: Me,
    underline: He,
    strong: De,
    code: Ue,
    italic: Be,
    link: ze,
    styled: Fe,
    subscript: qe,
    superscript: Ve,
    anchor: Ge,
    highlight: Ke,
    textStyle: Je
  }
}, We = function(r) {
  const e = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#39;"
  }, t = /[&<>"']/g, s = new RegExp(t.source);
  return r && s.test(r) ? r.replace(t, (o) => e[o]) : r;
};
let q = !1, N = class {
  constructor(r) {
    m(this, "marks"), m(this, "nodes"), r || (r = Ye), this.marks = r.marks || [], this.nodes = r.nodes || [];
  }
  addNode(r, e) {
    this.nodes[r] = e;
  }
  addMark(r, e) {
    this.marks[r] = e;
  }
  render(r, e = { optimizeImages: !1 }, t = !0) {
    if (!q && t && (console.warn(
      "Warning ⚠️: The RichTextResolver class is deprecated and will be removed in the next major release. Please use the `@storyblok/richtext` package instead. https://github.com/storyblok/richtext/"
    ), q = !0), r && r.content && Array.isArray(r.content)) {
      let s = "";
      return r.content.forEach((o) => {
        s += this.renderNode(o);
      }), e.optimizeImages ? this.optimizeImages(s, e.optimizeImages) : s;
    }
    return console.warn(
      `The render method must receive an Object with a "content" field.
      The "content" field must be an array of nodes as the type ISbRichtext.
      ISbRichtext:
        content?: ISbRichtext[]
        marks?: ISbRichtext[]
        attrs?: any
        text?: string
        type: string
        
        Example:
        {
          content: [
            {
              content: [
                {
                  text: 'Hello World',
                  type: 'text'
                }
              ],
              type: 'paragraph'
            }
          ],
          type: 'doc'
        }`
    ), "";
  }
  optimizeImages(r, e) {
    let t = 0, s = 0, o = "", n = "";
    typeof e != "boolean" && (typeof e.width == "number" && e.width > 0 && (o += `width="${e.width}" `, t = e.width), typeof e.height == "number" && e.height > 0 && (o += `height="${e.height}" `, s = e.height), (e.loading === "lazy" || e.loading === "eager") && (o += `loading="${e.loading}" `), typeof e.class == "string" && e.class.length > 0 && (o += `class="${e.class}" `), e.filters && (typeof e.filters.blur == "number" && e.filters.blur >= 0 && e.filters.blur <= 100 && (n += `:blur(${e.filters.blur})`), typeof e.filters.brightness == "number" && e.filters.brightness >= -100 && e.filters.brightness <= 100 && (n += `:brightness(${e.filters.brightness})`), e.filters.fill && (e.filters.fill.match(/[0-9A-F]{6}/gi) || e.filters.fill === "transparent") && (n += `:fill(${e.filters.fill})`), e.filters.format && ["webp", "png", "jpeg"].includes(e.filters.format) && (n += `:format(${e.filters.format})`), typeof e.filters.grayscale == "boolean" && e.filters.grayscale && (n += ":grayscale()"), typeof e.filters.quality == "number" && e.filters.quality >= 0 && e.filters.quality <= 100 && (n += `:quality(${e.filters.quality})`), e.filters.rotate && [90, 180, 270].includes(e.filters.rotate) && (n += `:rotate(${e.filters.rotate})`), n.length > 0 && (n = `/filters${n}`))), o.length > 0 && (r = r.replace(/<img/g, `<img ${o.trim()}`));
    const a = t > 0 || s > 0 || n.length > 0 ? `${t}x${s}${n}` : "";
    return r = r.replace(
      /a.storyblok.com\/f\/(\d+)\/([^.]+)\.(gif|jpg|jpeg|png|tif|bmp)/g,
      `a.storyblok.com/f/$1/$2.$3/m/${a}`
    ), typeof e != "boolean" && (e.sizes || e.srcset) && (r = r.replace(/<img.*?src=["|'](.*?)["|']/g, (i) => {
      var l, h;
      const u = i.match(
        /a.storyblok.com\/f\/(\d+)\/([^.]+)\.(gif|jpg|jpeg|png|tif|bmp)/g
      );
      if (u && u.length > 0) {
        const p = {
          srcset: (l = e.srcset) == null ? void 0 : l.map(($) => {
            if (typeof $ == "number")
              return `//${u}/m/${$}x0${n} ${$}w`;
            if (typeof $ == "object" && $.length === 2) {
              let T = 0, E = 0;
              return typeof $[0] == "number" && (T = $[0]), typeof $[1] == "number" && (E = $[1]), `//${u}/m/${T}x${E}${n} ${T}w`;
            }
            return "";
          }).join(", "),
          sizes: (h = e.sizes) == null ? void 0 : h.map(($) => $).join(", ")
        };
        let b = "";
        return p.srcset && (b += `srcset="${p.srcset}" `), p.sizes && (b += `sizes="${p.sizes}" `), i.replace(/<img/g, `<img ${b.trim()}`);
      }
      return i;
    })), r;
  }
  renderNode(r) {
    const e = [];
    r.marks && r.marks.forEach((s) => {
      const o = this.getMatchingMark(s);
      o && o.tag !== "" && e.push(this.renderOpeningTag(o.tag));
    });
    const t = this.getMatchingNode(r);
    return t && t.tag && e.push(this.renderOpeningTag(t.tag)), r.content ? r.content.forEach((s) => {
      e.push(this.renderNode(s));
    }) : r.text ? e.push(We(r.text)) : t && t.singleTag ? e.push(this.renderTag(t.singleTag, " /")) : t && t.html ? e.push(t.html) : r.type === "emoji" && e.push(this.renderEmoji(r)), t && t.tag && e.push(this.renderClosingTag(t.tag)), r.marks && r.marks.slice(0).reverse().forEach((s) => {
      const o = this.getMatchingMark(s);
      o && o.tag !== "" && e.push(this.renderClosingTag(o.tag));
    }), e.join("");
  }
  renderTag(r, e) {
    return r.constructor === String ? `<${r}${e}>` : r.map((t) => {
      if (t.constructor === String)
        return `<${t}${e}>`;
      {
        let s = `<${t.tag}`;
        if (t.attrs) {
          for (const o in t.attrs)
            if (Object.prototype.hasOwnProperty.call(t.attrs, o)) {
              const n = t.attrs[o];
              n !== null && (s += ` ${o}="${n}"`);
            }
        }
        return `${s}${e}>`;
      }
    }).join("");
  }
  renderOpeningTag(r) {
    return this.renderTag(r, "");
  }
  renderClosingTag(r) {
    return r.constructor === String ? `</${r}>` : r.slice(0).reverse().map((e) => e.constructor === String ? `</${e}>` : `</${e.tag}>`).join("");
  }
  getMatchingNode(r) {
    const e = this.nodes[r.type];
    if (typeof e == "function")
      return e(r);
  }
  getMatchingMark(r) {
    const e = this.marks[r.type];
    if (typeof e == "function")
      return e(r);
  }
  renderEmoji(r) {
    if (r.attrs.emoji)
      return r.attrs.emoji;
    const e = [
      {
        tag: "img",
        attrs: {
          src: r.attrs.fallbackImage,
          draggable: "false",
          loading: "lazy",
          align: "absmiddle"
        }
      }
    ];
    return this.renderTag(e, " /");
  }
};
class Xe {
  constructor(e) {
    m(this, "baseURL"), m(this, "timeout"), m(this, "headers"), m(this, "responseInterceptor"), m(this, "fetch"), m(this, "ejectInterceptor"), m(this, "url"), m(this, "parameters"), m(this, "fetchOptions"), this.baseURL = e.baseURL, this.headers = e.headers || new Headers(), this.timeout = e != null && e.timeout ? e.timeout * 1e3 : 0, this.responseInterceptor = e.responseInterceptor, this.fetch = (...t) => e.fetch ? e.fetch(...t) : fetch(...t), this.ejectInterceptor = !1, this.url = "", this.parameters = {}, this.fetchOptions = {};
  }
  /**
   *
   * @param url string
   * @param params ISbStoriesParams
   * @returns Promise<ISbResponse | Error>
   */
  get(e, t) {
    return this.url = e, this.parameters = t, this._methodHandler("get");
  }
  post(e, t) {
    return this.url = e, this.parameters = t, this._methodHandler("post");
  }
  put(e, t) {
    return this.url = e, this.parameters = t, this._methodHandler("put");
  }
  delete(e, t) {
    return this.url = e, this.parameters = t ?? {}, this._methodHandler("delete");
  }
  async _responseHandler(e) {
    const t = [], s = {
      data: {},
      headers: {},
      status: 0,
      statusText: ""
    };
    e.status !== 204 && await e.json().then((o) => {
      s.data = o;
    });
    for (const o of e.headers.entries())
      t[o[0]] = o[1];
    return s.headers = { ...t }, s.status = e.status, s.statusText = e.statusText, s;
  }
  async _methodHandler(e) {
    let t = `${this.baseURL}${this.url}`, s = null;
    if (e === "get") {
      const l = new P();
      t = `${this.baseURL}${this.url}?${l.stringify(
        this.parameters
      )}`;
    } else
      s = JSON.stringify(this.parameters);
    const o = new URL(t), n = new AbortController(), { signal: a } = n;
    let i;
    this.timeout && (i = setTimeout(() => n.abort(), this.timeout));
    try {
      const l = await this.fetch(`${o}`, {
        method: e,
        headers: this.headers,
        body: s,
        signal: a,
        ...this.fetchOptions
      });
      this.timeout && clearTimeout(i);
      const h = await this._responseHandler(
        l
      );
      return this.responseInterceptor && !this.ejectInterceptor ? this._statusHandler(this.responseInterceptor(h)) : this._statusHandler(h);
    } catch (l) {
      return {
        message: l
      };
    }
  }
  setFetchOptions(e = {}) {
    Object.keys(e).length > 0 && "method" in e && delete e.method, this.fetchOptions = { ...e };
  }
  eject() {
    this.ejectInterceptor = !0;
  }
  _statusHandler(e) {
    const t = /20[0-6]/g;
    return new Promise((s, o) => {
      if (t.test(`${e.status}`))
        return s(e);
      const n = {
        message: e.statusText,
        status: e.status,
        response: Array.isArray(e.data) ? e.data[0] : e.data.error || e.data.slug
      };
      o(n);
    });
  }
}
const V = "SB-Agent", D = {
  defaultAgentName: "SB-JS-CLIENT",
  defaultAgentVersion: "SB-Agent-Version",
  packageVersion: "6.0.0"
};
let C = {};
const x = {};
class Qe {
  /**
   *
   * @param config ISbConfig interface
   * @param pEndpoint string, optional
   */
  constructor(e, t) {
    m(this, "client"), m(this, "maxRetries"), m(this, "retriesDelay"), m(this, "throttle"), m(this, "accessToken"), m(this, "cache"), m(this, "helpers"), m(this, "resolveCounter"), m(this, "relations"), m(this, "links"), m(this, "richTextResolver"), m(this, "resolveNestedRelations"), m(this, "stringifiedStoriesCache");
    let s = e.endpoint || t;
    if (!s) {
      const a = new P().getRegionURL, i = e.https === !1 ? "http" : "https";
      e.oauthToken ? s = `${i}://${a(e.region)}/v1` : s = `${i}://${a(e.region)}/v2`;
    }
    const o = new Headers();
    o.set("Content-Type", "application/json"), o.set("Accept", "application/json"), e.headers && (e.headers.constructor.name === "Headers" ? e.headers.entries().toArray() : Object.entries(e.headers)).forEach(([a, i]) => {
      o.set(a, i);
    }), o.has(V) || (o.set(V, D.defaultAgentName), o.set(
      D.defaultAgentVersion,
      D.packageVersion
    ));
    let n = 5;
    e.oauthToken && (o.set("Authorization", e.oauthToken), n = 3), e.rateLimit && (n = e.rateLimit), e.richTextSchema ? this.richTextResolver = new N(e.richTextSchema) : this.richTextResolver = new N(), e.componentResolver && this.setComponentResolver(e.componentResolver), this.maxRetries = e.maxRetries || 10, this.retriesDelay = 300, this.throttle = Te(
      this.throttledRequest.bind(this),
      n,
      1e3
    ), this.accessToken = e.accessToken || "", this.relations = {}, this.links = {}, this.cache = e.cache || { clear: "manual" }, this.helpers = new P(), this.resolveCounter = 0, this.resolveNestedRelations = e.resolveNestedRelations || !0, this.stringifiedStoriesCache = {}, this.client = new Xe({
      baseURL: s,
      timeout: e.timeout || 0,
      headers: o,
      responseInterceptor: e.responseInterceptor,
      fetch: e.fetch
    });
  }
  setComponentResolver(e) {
    this.richTextResolver.addNode("blok", (t) => {
      let s = "";
      return t.attrs.body && t.attrs.body.forEach((o) => {
        s += e(o.component, o);
      }), {
        html: s
      };
    });
  }
  parseParams(e) {
    return e.token || (e.token = this.getToken()), e.cv || (e.cv = x[e.token]), Array.isArray(e.resolve_relations) && (e.resolve_relations = e.resolve_relations.join(",")), typeof e.resolve_relations < "u" && (e.resolve_level = 2), e;
  }
  factoryParamOptions(e, t) {
    return this.helpers.isCDNUrl(e) ? this.parseParams(t) : t;
  }
  makeRequest(e, t, s, o, n) {
    const a = this.factoryParamOptions(
      e,
      this.helpers.getOptionsPage(t, s, o)
    );
    return this.cacheResponse(e, a, void 0, n);
  }
  get(e, t, s) {
    t || (t = {});
    const o = `/${e}`, n = this.factoryParamOptions(o, t);
    return this.cacheResponse(o, n, void 0, s);
  }
  async getAll(e, t, s, o) {
    const n = (t == null ? void 0 : t.per_page) || 25, a = `/${e}`.replace(/\/$/, ""), i = s ?? a.substring(a.lastIndexOf("/") + 1), l = 1, h = await this.makeRequest(
      a,
      t,
      n,
      l,
      o
    ), u = h.total ? Math.ceil(h.total / n) : 1, p = await this.helpers.asyncMap(
      this.helpers.range(l, u),
      (b) => this.makeRequest(a, t, n, b + 1, o)
    );
    return this.helpers.flatMap([h, ...p], (b) => Object.values(b.data[i]));
  }
  post(e, t, s) {
    const o = `/${e}`;
    return Promise.resolve(
      this.throttle("post", o, t, s)
    );
  }
  put(e, t, s) {
    const o = `/${e}`;
    return Promise.resolve(
      this.throttle("put", o, t, s)
    );
  }
  delete(e, t, s) {
    t || (t = {});
    const o = `/${e}`;
    return Promise.resolve(
      this.throttle("delete", o, t, s)
    );
  }
  getStories(e, t) {
    return this._addResolveLevel(e), this.get("cdn/stories", e, t);
  }
  getStory(e, t, s) {
    return this._addResolveLevel(t), this.get(`cdn/stories/${e}`, t, s);
  }
  getToken() {
    return this.accessToken;
  }
  ejectInterceptor() {
    this.client.eject();
  }
  _addResolveLevel(e) {
    typeof e.resolve_relations < "u" && (e.resolve_level = 2);
  }
  _cleanCopy(e) {
    return JSON.parse(JSON.stringify(e));
  }
  _insertLinks(e, t, s) {
    const o = e[t];
    o && o.fieldtype === "multilink" && o.linktype === "story" && typeof o.id == "string" && this.links[s][o.id] ? o.story = this._cleanCopy(this.links[s][o.id]) : o && o.linktype === "story" && typeof o.uuid == "string" && this.links[s][o.uuid] && (o.story = this._cleanCopy(this.links[s][o.uuid]));
  }
  /**
   *
   * @param resolveId A counter number as a string
   * @param uuid The uuid of the story
   * @returns string | object
   */
  getStoryReference(e, t) {
    return this.relations[e][t] ? JSON.parse(this.stringifiedStoriesCache[t] || JSON.stringify(this.relations[e][t])) : t;
  }
  /**
   * Resolves a field's value by replacing UUIDs with their corresponding story references
   * @param jtree - The JSON tree object containing the field to resolve
   * @param treeItem - The key of the field to resolve
   * @param resolveId - The unique identifier for the current resolution context
   *
   * This method handles both single string UUIDs and arrays of UUIDs:
   * - For single strings: directly replaces the UUID with the story reference
   * - For arrays: maps through each UUID and replaces with corresponding story references
   */
  _resolveField(e, t, s) {
    const o = e[t];
    typeof o == "string" ? e[t] = this.getStoryReference(s, o) : Array.isArray(o) && (e[t] = o.map(
      (n) => this.getStoryReference(s, n)
    ).filter(Boolean));
  }
  /**
   * Inserts relations into the JSON tree by resolving references
   * @param jtree - The JSON tree object to process
   * @param treeItem - The current field being processed
   * @param fields - The relation patterns to resolve (string or array of strings)
   * @param resolveId - The unique identifier for the current resolution context
   *
   * This method handles two types of relation patterns:
   * 1. Nested relations: matches fields that end with the current field name
   *    Example: If treeItem is "event_type", it matches patterns like "*.event_type"
   *
   * 2. Direct component relations: matches exact component.field patterns
   *    Example: "event.event_type" for component "event" and field "event_type"
   *
   * The method supports both string and array formats for the fields parameter,
   * allowing flexible specification of relation patterns.
   */
  _insertRelations(e, t, s, o) {
    if (Array.isArray(s) ? s.find((a) => a.endsWith(`.${t}`)) : s.endsWith(`.${t}`)) {
      this._resolveField(e, t, o);
      return;
    }
    const n = e.component ? `${e.component}.${t}` : t;
    (Array.isArray(s) ? s.includes(n) : s === n) && this._resolveField(e, t, o);
  }
  /**
   * Recursively traverses and resolves relations in the story content tree
   * @param story - The story object containing the content to process
   * @param fields - The relation patterns to resolve
   * @param resolveId - The unique identifier for the current resolution context
   */
  iterateTree(e, t, s) {
    const o = (n, a = "") => {
      if (!(!n || n._stopResolving)) {
        if (Array.isArray(n))
          n.forEach((i, l) => o(i, `${a}[${l}]`));
        else if (typeof n == "object")
          for (const i in n) {
            const l = a ? `${a}.${i}` : i;
            (n.component && n._uid || n.type === "link") && (this._insertRelations(n, i, t, s), this._insertLinks(n, i, s)), o(n[i], l);
          }
      }
    };
    o(e.content);
  }
  async resolveLinks(e, t, s) {
    let o = [];
    if (e.link_uuids) {
      const n = e.link_uuids.length, a = [], i = 50;
      for (let l = 0; l < n; l += i) {
        const h = Math.min(n, l + i);
        a.push(e.link_uuids.slice(l, h));
      }
      for (let l = 0; l < a.length; l++)
        (await this.getStories({
          per_page: i,
          language: t.language,
          version: t.version,
          starts_with: t.starts_with,
          by_uuids: a[l].join(",")
        })).data.stories.forEach(
          (h) => {
            o.push(h);
          }
        );
    } else
      o = e.links;
    o.forEach((n) => {
      this.links[s][n.uuid] = {
        ...n,
        _stopResolving: !0
      };
    });
  }
  async resolveRelations(e, t, s) {
    let o = [];
    if (e.rel_uuids) {
      const n = e.rel_uuids.length, a = [], i = 50;
      for (let l = 0; l < n; l += i) {
        const h = Math.min(n, l + i);
        a.push(e.rel_uuids.slice(l, h));
      }
      for (let l = 0; l < a.length; l++)
        (await this.getStories({
          per_page: i,
          language: t.language,
          version: t.version,
          starts_with: t.starts_with,
          by_uuids: a[l].join(","),
          excluding_fields: t.excluding_fields
        })).data.stories.forEach((h) => {
          o.push(h);
        });
    } else
      o = e.rels;
    o && o.length > 0 && o.forEach((n) => {
      this.relations[s][n.uuid] = {
        ...n,
        _stopResolving: !0
      };
    });
  }
  /**
   *
   * @param responseData
   * @param params
   * @param resolveId
   * @description Resolves the relations and links of the stories
   * @returns Promise<void>
   *
   */
  async resolveStories(e, t, s) {
    var o, n;
    let a = [];
    if (this.links[s] = {}, this.relations[s] = {}, typeof t.resolve_relations < "u" && t.resolve_relations.length > 0 && (typeof t.resolve_relations == "string" && (a = t.resolve_relations.split(",")), await this.resolveRelations(e, t, s)), t.resolve_links && ["1", "story", "url", "link"].includes(t.resolve_links) && ((o = e.links) != null && o.length || (n = e.link_uuids) != null && n.length) && await this.resolveLinks(e, t, s), this.resolveNestedRelations)
      for (const i in this.relations[s])
        this.iterateTree(
          this.relations[s][i],
          a,
          s
        );
    e.story ? this.iterateTree(e.story, a, s) : e.stories.forEach((i) => {
      this.iterateTree(i, a, s);
    }), this.stringifiedStoriesCache = {}, delete this.links[s], delete this.relations[s];
  }
  async cacheResponse(e, t, s, o) {
    const n = this.helpers.stringify({ url: e, params: t }), a = this.cacheProvider();
    if (t.version === "published" && e !== "/cdn/spaces/me") {
      const i = await a.get(n);
      if (i)
        return Promise.resolve(i);
    }
    return new Promise(async (i, l) => {
      var h;
      try {
        const u = await this.throttle(
          "get",
          e,
          t,
          o
        );
        if (u.status !== 200)
          return l(u);
        let p = { data: u.data, headers: u.headers };
        if ((h = u.headers) != null && h["per-page"] && (p = Object.assign({}, p, {
          perPage: u.headers["per-page"] ? Number.parseInt(u.headers["per-page"]) : 0,
          total: u.headers["per-page"] ? Number.parseInt(u.headers.total) : 0
        })), p.data.story || p.data.stories) {
          const $ = this.resolveCounter = ++this.resolveCounter % 1e3;
          await this.resolveStories(p.data, t, `${$}`);
        }
        t.version === "published" && e !== "/cdn/spaces/me" && await a.set(n, p);
        const b = this.cache.clear === "onpreview" && t.version === "draft" || this.cache.clear === "auto";
        return t.token && p.data.cv && (b && x[t.token] && x[t.token] !== p.data.cv && await this.flushCache(), x[t.token] = p.data.cv), i(p);
      } catch (u) {
        if (u.response && u.status === 429 && (s = typeof s > "u" ? 0 : s + 1, s < this.maxRetries))
          return console.log(
            `Hit rate limit. Retrying in ${this.retriesDelay / 1e3} seconds.`
          ), await this.helpers.delay(this.retriesDelay), this.cacheResponse(e, t, s).then(i).catch(l);
        l(u);
      }
    });
  }
  throttledRequest(e, t, s, o) {
    return this.client.setFetchOptions(o), this.client[e](t, s);
  }
  cacheVersions() {
    return x;
  }
  cacheVersion() {
    return x[this.accessToken];
  }
  setCacheVersion(e) {
    this.accessToken && (x[this.accessToken] = e);
  }
  clearCacheVersion() {
    this.accessToken && (x[this.accessToken] = 0);
  }
  cacheProvider() {
    switch (this.cache.type) {
      case "memory":
        return {
          get(e) {
            return Promise.resolve(C[e]);
          },
          getAll() {
            return Promise.resolve(C);
          },
          set(e, t) {
            return C[e] = t, Promise.resolve(void 0);
          },
          flush() {
            return C = {}, Promise.resolve(void 0);
          }
        };
      case "custom":
        if (this.cache.custom)
          return this.cache.custom;
      // eslint-disable-next-line no-fallthrough
      default:
        return {
          get() {
            return Promise.resolve();
          },
          getAll() {
            return Promise.resolve(void 0);
          },
          set() {
            return Promise.resolve(void 0);
          },
          flush() {
            return Promise.resolve(void 0);
          }
        };
    }
  }
  async flushCache() {
    return await this.cacheProvider().flush(), this.clearCacheVersion(), this;
  }
}
const gt = (r = {}) => {
  const { apiOptions: e } = r;
  if (!e || !e.accessToken) {
    console.error(
      "You need to provide an access token to interact with Storyblok API. Read https://www.storyblok.com/docs/api/content-delivery#topics/authentication"
    );
    return;
  }
  return { storyblokApi: new Qe(e) };
}, Ze = (r) => {
  if (typeof r != "object" || typeof r._editable > "u")
    return {};
  try {
    const e = JSON.parse(
      r._editable.replace(/^<!--#storyblok#/, "").replace(/-->$/, "")
    );
    return e ? {
      "data-blok-c": JSON.stringify(e),
      "data-blok-uid": `${e.id}-${e.uid}`
    } : {};
  } catch {
    return {};
  }
};
function et(r, e) {
  if (!e)
    return { src: r, attrs: {} };
  let t = 0, s = 0;
  const o = {}, n = [];
  function a(l, h, u, p, b) {
    typeof l != "number" || l <= h || l >= u ? console.warn(`[StoryblokRichText] - ${p.charAt(0).toUpperCase() + p.slice(1)} value must be a number between ${h} and ${u} (inclusive)`) : b.push(`${p}(${l})`);
  }
  if (typeof e == "object") {
    if (typeof e.width == "number" && e.width > 0 ? (o.width = e.width, t = e.width) : console.warn("[StoryblokRichText] - Width value must be a number greater than 0"), e.height && typeof e.height == "number" && e.height > 0 ? (o.height = e.height, s = e.height) : console.warn("[StoryblokRichText] - Height value must be a number greater than 0"), e.loading && ["lazy", "eager"].includes(e.loading) && (o.loading = e.loading), e.class && (o.class = e.class), e.filters) {
      const { filters: l } = e || {}, { blur: h, brightness: u, fill: p, format: b, grayscale: $, quality: T, rotate: E } = l || {};
      h && a(h, 0, 100, "blur", n), T && a(T, 0, 100, "quality", n), u && a(u, 0, 100, "brightness", n), p && n.push(`fill(${p})`), $ && n.push("grayscale()"), E && [0, 90, 180, 270].includes(e.filters.rotate || 0) && n.push(`rotate(${E})`), b && ["webp", "png", "jpeg"].includes(b) && n.push(`format(${b})`);
    }
    e.srcset && (o.srcset = e.srcset.map((l) => {
      if (typeof l == "number")
        return `${r}/m/${l}x0/${n.length > 0 ? `filters:${n.join(":")}` : ""} ${l}w`;
      if (Array.isArray(l) && l.length === 2) {
        const [h, u] = l;
        return `${r}/m/${h}x${u}/${n.length > 0 ? `filters:${n.join(":")}` : ""} ${h}w`;
      } else {
        console.warn("[StoryblokRichText] - srcset entry must be a number or a tuple of two numbers");
        return;
      }
    }).join(", ")), e.sizes && (o.sizes = e.sizes.join(", "));
  }
  let i = `${r}/m/`;
  return t > 0 && s > 0 && (i = `${i}${t}x${s}/`), n.length > 0 && (i = `${i}filters:${n.join(":")}`), {
    src: i,
    attrs: o
  };
}
var k = /* @__PURE__ */ ((r) => (r.DOCUMENT = "doc", r.HEADING = "heading", r.PARAGRAPH = "paragraph", r.QUOTE = "blockquote", r.OL_LIST = "ordered_list", r.UL_LIST = "bullet_list", r.LIST_ITEM = "list_item", r.CODE_BLOCK = "code_block", r.HR = "horizontal_rule", r.BR = "hard_break", r.IMAGE = "image", r.EMOJI = "emoji", r.COMPONENT = "blok", r.TABLE = "table", r.TABLE_ROW = "tableRow", r.TABLE_CELL = "tableCell", r.TABLE_HEADER = "tableHeader", r))(k || {}), S = /* @__PURE__ */ ((r) => (r.BOLD = "bold", r.STRONG = "strong", r.STRIKE = "strike", r.UNDERLINE = "underline", r.ITALIC = "italic", r.CODE = "code", r.LINK = "link", r.ANCHOR = "anchor", r.STYLED = "styled", r.SUPERSCRIPT = "superscript", r.SUBSCRIPT = "subscript", r.TEXT_STYLE = "textStyle", r.HIGHLIGHT = "highlight", r))(S || {}), Q = /* @__PURE__ */ ((r) => (r.TEXT = "text", r))(Q || {}), j = /* @__PURE__ */ ((r) => (r.URL = "url", r.STORY = "story", r.ASSET = "asset", r.EMAIL = "email", r))(j || {});
const tt = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
], rt = (r = {}) => Object.keys(r).map((e) => `${e}="${r[e]}"`).join(" "), st = (r = {}) => Object.keys(r).map((e) => `${e}: ${r[e]}`).join("; ");
function ot(r) {
  return r.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
const L = (r) => Object.fromEntries(Object.entries(r).filter(([e, t]) => t !== void 0));
function G(r, e = {}, t) {
  const s = rt(e), o = s ? `${r} ${s}` : r, n = Array.isArray(t) ? t.join("") : t || "";
  if (r) {
    if (tt.includes(r))
      return `<${o}>`;
  } else return n;
  return `<${o}>${n}</${r}>`;
}
function nt(r = {}) {
  const e = /* @__PURE__ */ new Map(), {
    renderFn: t = G,
    textFn: s = ot,
    resolvers: o = {},
    optimizeImages: n = !1,
    keyedResolvers: a = !1
  } = r, i = t !== G, l = () => ({ render: (c, d = {}, g) => {
    if (a && c) {
      const f = e.get(c) || 0;
      e.set(c, f + 1), d.key = `${c}-${f}`;
    }
    return t(c, d, g);
  } }), h = (c) => (d, g) => {
    const f = d.attrs || {};
    return g.render(c, f, d.children || null);
  }, u = (c, d) => {
    const { src: g, alt: f, title: y, srcset: R, sizes: v } = c.attrs || {};
    let w = g, _ = {};
    if (n) {
      const { src: he, attrs: ue } = et(g, n);
      w = he, _ = ue;
    }
    const ce = {
      src: w,
      alt: f,
      title: y,
      srcset: R,
      sizes: v,
      ..._
    };
    return d.render("img", L(ce));
  }, p = (c, d) => {
    const { level: g, ...f } = c.attrs || {};
    return d.render(`h${g}`, f, c.children);
  }, b = (c, d) => {
    var g, f, y, R;
    const v = d.render("img", {
      src: (g = c.attrs) == null ? void 0 : g.fallbackImage,
      alt: (f = c.attrs) == null ? void 0 : f.alt,
      style: "width: 1.25em; height: 1.25em; vertical-align: text-top",
      draggable: "false",
      loading: "lazy"
    });
    return d.render("span", {
      "data-type": "emoji",
      "data-name": (y = c.attrs) == null ? void 0 : y.name,
      "data-emoji": (R = c.attrs) == null ? void 0 : R.emoji
    }, v);
  }, $ = (c, d) => d.render(
    "pre",
    c.attrs || {},
    d.render("code", {}, c.children || "")
  ), T = (c, d = !1) => ({ text: g, attrs: f }, y) => {
    const { class: R, id: v, ...w } = f || {}, _ = d ? {
      class: R,
      id: v,
      style: st(w) || void 0
    } : f || {};
    return y.render(c, L(_), g);
  }, E = (c) => M(c), re = (c) => {
    const { marks: d, ...g } = c;
    if ("text" in c) {
      if (d)
        return d.reduce(
          (y, R) => E({ ...R, text: y }),
          E({ ...g, children: g.children })
        );
      const f = c.attrs || {};
      if (a) {
        const y = e.get("txt") || 0;
        e.set("txt", y + 1), f.key = `txt-${y}`;
      }
      return s(g.text, f);
    }
    return "";
  }, B = (c, d) => {
    const { linktype: g, href: f, anchor: y, ...R } = c.attrs || {};
    let v = "";
    switch (g) {
      case j.ASSET:
      case j.URL:
        v = f;
        break;
      case j.EMAIL:
        v = `mailto:${f}`;
        break;
      case j.STORY:
        v = f, y && (v = `${v}#${y}`);
        break;
      default:
        v = f;
        break;
    }
    const w = { ...R };
    return v && (w.href = v), d.render("a", w, c.text);
  }, se = (c, d) => {
    var g, f;
    return console.warn("[StoryblokRichtText] - BLOK resolver is not available for vanilla usage"), d.render("span", {
      blok: (g = c == null ? void 0 : c.attrs) == null ? void 0 : g.body[0],
      id: (f = c.attrs) == null ? void 0 : f.id,
      style: "display: none"
    });
  }, oe = (c, d) => {
    const g = {}, f = d.render("tbody", {}, c.children);
    return d.render("table", g, f);
  }, ne = (c, d) => {
    const g = {};
    return d.render("tr", g, c.children);
  }, ae = (c, d) => {
    const { colspan: g, rowspan: f, colwidth: y, backgroundColor: R, ...v } = c.attrs || {}, w = {
      ...v
    };
    g > 1 && (w.colspan = g), f > 1 && (w.rowspan = f);
    const _ = [];
    return y && _.push(`width: ${y}px;`), R && _.push(`background-color: ${R};`), _.length > 0 && (w.style = _.join(" ")), d.render("td", L(w), c.children);
  }, ie = (c, d) => {
    const { colspan: g, rowspan: f, colwidth: y, backgroundColor: R, ...v } = c.attrs || {}, w = {
      ...v
    };
    g > 1 && (w.colspan = g), f > 1 && (w.rowspan = f);
    const _ = [];
    return y && _.push(`width: ${y}px;`), R && _.push(`background-color: ${R};`), _.length > 0 && (w.style = _.join(" ")), d.render("th", L(w), c.children);
  }, le = new Map([
    [k.DOCUMENT, h("")],
    [k.HEADING, p],
    [k.PARAGRAPH, h("p")],
    [k.UL_LIST, h("ul")],
    [k.OL_LIST, h("ol")],
    [k.LIST_ITEM, h("li")],
    [k.IMAGE, u],
    [k.EMOJI, b],
    [k.CODE_BLOCK, $],
    [k.HR, h("hr")],
    [k.BR, h("br")],
    [k.QUOTE, h("blockquote")],
    [k.COMPONENT, se],
    [Q.TEXT, re],
    [S.LINK, B],
    [S.ANCHOR, B],
    [S.STYLED, T("span", !0)],
    [S.BOLD, T("strong")],
    [S.TEXT_STYLE, T("span", !0)],
    [S.ITALIC, T("em")],
    [S.UNDERLINE, T("u")],
    [S.STRIKE, T("s")],
    [S.CODE, T("code")],
    [S.SUPERSCRIPT, T("sup")],
    [S.SUBSCRIPT, T("sub")],
    [S.HIGHLIGHT, T("mark")],
    [k.TABLE, oe],
    [k.TABLE_ROW, ne],
    [k.TABLE_CELL, ae],
    [k.TABLE_HEADER, ie],
    ...Object.entries(o).map(([c, d]) => [c, d])
  ]);
  function I(c) {
    const d = le.get(c.type);
    if (!d)
      return console.error("<Storyblok>", `No resolver found for node type ${c.type}`), "";
    const g = l();
    if (c.type === "text")
      return d(c, g);
    const f = c.content ? c.content.map(M) : void 0;
    return d({
      ...c,
      children: f
    }, g);
  }
  function M(c) {
    return c.type === "doc" ? i ? c.content.map(I) : c.content.map(I).join("") : Array.isArray(c) ? c.map(I) : I(c);
  }
  return {
    render: M
  };
}
let U, K = "https://app.storyblok.com/f/storyblok-v2-latest.js";
const at = (r, e, t = {}) => {
  var s;
  const o = !(typeof window > "u") && typeof window.storyblokRegisterEvent < "u", n = new URL((s = window.location) == null ? void 0 : s.href).searchParams.get(
    "_storyblok"
  ), a = n !== null && +n === r;
  if (!(!o || !a)) {
    if (!r) {
      console.warn("Story ID is not defined. Please provide a valid ID.");
      return;
    }
    window.storyblokRegisterEvent(() => {
      new window.StoryblokBridge(t).on(["input", "published", "change"], (i) => {
        var l;
        i && (i.action === "input" && ((l = i.story) == null ? void 0 : l.id) === r ? e(i.story) : (i.action === "change" || i.action === "published") && i.storyId === r && window.location.reload());
      });
    });
  }
}, Z = (r, e) => {
  r.addNode("blok", (t) => {
    let s = "";
    return t.attrs.body.forEach((o) => {
      s += e(o.component, o);
    }), {
      html: s
    };
  });
}, it = (r = {}) => {
  var e, t;
  const {
    bridge: s,
    accessToken: o,
    use: n = [],
    apiOptions: a = {},
    richText: i = {},
    bridgeUrl: l
  } = r;
  a.accessToken = a.accessToken || o;
  const h = { bridge: s, apiOptions: a };
  let u = {};
  n.forEach((b) => {
    u = { ...u, ...b(h) };
  }), l && (K = l);
  const p = !(typeof window > "u") && ((t = (e = window.location) == null ? void 0 : e.search) == null ? void 0 : t.includes("_storyblok_tk"));
  return s !== !1 && p && be(K), U = new N(i.schema), i.resolver && Z(U, i.resolver), u;
}, lt = (r) => {
  var e;
  return !r || !((e = r == null ? void 0 : r.content) != null && e.some((t) => t.content || t.type === "blok" || t.type === "horizontal_rule"));
}, ft = (r, e, t) => {
  let s = t || U;
  if (!s) {
    console.error(
      "Please initialize the Storyblok SDK before calling the renderRichText function"
    );
    return;
  }
  return lt(r) ? "" : (e && (s = new N(e.schema), e.resolver && Z(s, e.resolver)), s.render(r, {}, !1));
}, ee = /* @__PURE__ */ J({
  __name: "StoryblokComponent",
  props: {
    blok: {}
  },
  setup(r, { expose: e }) {
    const t = r, s = O();
    e({
      value: s
    });
    const o = typeof H(t.blok.component) != "string", n = de("VueSDKOptions"), a = O(t.blok.component);
    return !o && n && (n.enableFallbackComponent ? (a.value = n.customFallbackComponent ?? "FallbackComponent", typeof H(a.value) == "string" && console.error(
      `Is the Fallback component "${a.value}" registered properly?`
    )) : console.error(
      `Component could not be found for blok "${t.blok.component}"! Is it defined in main.ts as "app.component("${t.blok.component}", ${t.blok.component});"?`
    )), (i, l) => (W(), Y(H(a.value), pe({
      ref_key: "blokRef",
      ref: s
    }, { ...i.$props, ...i.$attrs }), null, 16));
  }
}), ct = (r) => {
  var e, t;
  return X(
    ee,
    {
      blok: (e = r == null ? void 0 : r.attrs) == null ? void 0 : e.body[0],
      id: (t = r.attrs) == null ? void 0 : t.id
    },
    r.children
  );
};
function ht(r) {
  const e = {
    renderFn: X,
    textFn: ge,
    keyedResolvers: !0,
    resolvers: {
      [k.COMPONENT]: ct,
      ...r.resolvers
    }
  };
  return nt(e);
}
const ut = /* @__PURE__ */ J({
  __name: "StoryblokRichText",
  props: {
    doc: {},
    resolvers: {}
  },
  setup(r) {
    const e = r, t = O(), s = () => t.value;
    return fe([() => e.doc, () => e.resolvers], ([o, n]) => {
      const { render: a } = ht({
        resolvers: n ?? {}
      });
      t.value = a(o);
    }, {
      immediate: !0,
      deep: !0
    }), (o, n) => (W(), Y(s));
  }
}), dt = {
  beforeMount(r, e) {
    if (e.value) {
      const t = Ze(e.value);
      Object.keys(t).length > 0 && (r.setAttribute("data-blok-c", t["data-blok-c"]), r.setAttribute("data-blok-uid", t["data-blok-uid"]), r.classList.add("storyblok__outline"));
    }
  }
}, te = (r) => {
  console.error(`You can't use ${r} if you're not loading apiPlugin. Please provide it on StoryblokVue initialization.
    `);
};
let A = null;
const mt = () => (A || te("useStoryblokApi"), A), yt = async (r, e = {}, t = {}) => {
  const s = O(null);
  if (t.resolveRelations = t.resolveRelations ?? e.resolve_relations, t.resolveLinks = t.resolveLinks ?? e.resolve_links, me(() => {
    s.value && s.value.id && at(
      s.value.id,
      (o) => s.value = o,
      t
    );
  }), A) {
    const { data: o } = await A.get(
      `cdn/stories/${r}`,
      e
    );
    s.value = o.story;
  } else
    te("useStoryblok");
  return s;
}, bt = {
  install(r, e = {}) {
    r.directive("editable", dt), r.component("StoryblokComponent", ee), r.component("StoryblokRichText", ut), e.enableFallbackComponent && !e.customFallbackComponent && r.component(
      "FallbackComponent",
      ye(() => import("./FallbackComponent-Dky11gEu.js"))
    );
    const { storyblokApi: t } = it(e);
    A = t || null, r.provide("VueSDKOptions", e);
  }
};
export {
  k as BlockTypes,
  S as MarkTypes,
  N as RichTextResolver,
  Ye as RichTextSchema,
  ee as StoryblokComponent,
  ut as StoryblokRichText,
  bt as StoryblokVue,
  Q as TextTypes,
  gt as apiPlugin,
  ft as renderRichText,
  nt as richTextResolver,
  yt as useStoryblok,
  mt as useStoryblokApi,
  at as useStoryblokBridge,
  ht as useStoryblokRichText
};
